{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "3217c063-399d-435b-aadd-f5cbe8270725",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "eb9c4c72-6ca7-4198-ad5c-0b2e533f4609",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\eisan'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "os.getcwd()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "82013cf2-14be-45c8-9dcf-f53b98b6962f",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "db583eca-fb2a-49c7-9941-0c43807cb3b0",
   "metadata": {},
   "outputs": [],
   "source": [
    "csv_path = r\"C:\\Users\\eisan\\PycharmProjects\\AMR first\\data\\Klebsiella_Meropenem_AMR_Genomics.csv\"\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "0afc877a-01bf-4699-a5a7-29626ee6fcf6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load Dataset \n",
    "df = pd.read_csv(csv_path)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "e3bf2914-d0ca-4e73-85f3-026752c79cc1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Shape: (1580, 302)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome ID</th>\n",
       "      <th>AAC(3)-IId</th>\n",
       "      <th>AAC(3)-IIe</th>\n",
       "      <th>AAC(3)-IV</th>\n",
       "      <th>AAC(3)-Ia</th>\n",
       "      <th>AAC(6')-29a</th>\n",
       "      <th>AAC(6')-33</th>\n",
       "      <th>AAC(6')-IIa</th>\n",
       "      <th>AAC(6')-IIc</th>\n",
       "      <th>AAC(6')-Ib-cr</th>\n",
       "      <th>...</th>\n",
       "      <th>Testing Standard</th>\n",
       "      <th>Testing Standard Year</th>\n",
       "      <th>Computational Method</th>\n",
       "      <th>Computational Method Version</th>\n",
       "      <th>Computational Method Performance</th>\n",
       "      <th>Evidence</th>\n",
       "      <th>Source</th>\n",
       "      <th>PubMed</th>\n",
       "      <th>SEQUENCE</th>\n",
       "      <th>Resistant Phenotype</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>573.13031</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>CLSI</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Laboratory Method</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28512093;28776045</td>\n",
       "      <td>573.13031</td>\n",
       "      <td>Susceptible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>573.13032</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>...</td>\n",
       "      <td>CLSI</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Laboratory Method</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28512093;28776045</td>\n",
       "      <td>573.13032</td>\n",
       "      <td>Susceptible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>573.13033</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>CLSI</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Laboratory Method</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28512093;28776045</td>\n",
       "      <td>573.13033</td>\n",
       "      <td>Susceptible</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>573.13034</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>CLSI</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Laboratory Method</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28512093;28776045</td>\n",
       "      <td>573.13034</td>\n",
       "      <td>Resistant</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>573.13035</td>\n",
       "      <td>0</td>\n",
       "      <td>1</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>...</td>\n",
       "      <td>CLSI</td>\n",
       "      <td>2017.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Laboratory Method</td>\n",
       "      <td>NaN</td>\n",
       "      <td>28512093;28776045</td>\n",
       "      <td>573.13035</td>\n",
       "      <td>Susceptible</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows Ã— 302 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   Genome ID  AAC(3)-IId  AAC(3)-IIe  AAC(3)-IV  AAC(3)-Ia  AAC(6')-29a  \\\n",
       "0  573.13031           0           1          0          0            0   \n",
       "1  573.13032           0           1          0          0            0   \n",
       "2  573.13033           0           1          0          0            0   \n",
       "3  573.13034           0           0          0          0            0   \n",
       "4  573.13035           0           1          0          0            0   \n",
       "\n",
       "   AAC(6')-33  AAC(6')-IIa  AAC(6')-IIc  AAC(6')-Ib-cr  ...  Testing Standard  \\\n",
       "0           0            0            0              1  ...              CLSI   \n",
       "1           0            0            0              1  ...              CLSI   \n",
       "2           0            0            0              0  ...              CLSI   \n",
       "3           0            0            0              0  ...              CLSI   \n",
       "4           0            0            0              0  ...              CLSI   \n",
       "\n",
       "   Testing Standard Year  Computational Method  Computational Method Version  \\\n",
       "0                 2017.0                   NaN                           NaN   \n",
       "1                 2017.0                   NaN                           NaN   \n",
       "2                 2017.0                   NaN                           NaN   \n",
       "3                 2017.0                   NaN                           NaN   \n",
       "4                 2017.0                   NaN                           NaN   \n",
       "\n",
       "   Computational Method Performance           Evidence  Source  \\\n",
       "0                               NaN  Laboratory Method     NaN   \n",
       "1                               NaN  Laboratory Method     NaN   \n",
       "2                               NaN  Laboratory Method     NaN   \n",
       "3                               NaN  Laboratory Method     NaN   \n",
       "4                               NaN  Laboratory Method     NaN   \n",
       "\n",
       "              PubMed   SEQUENCE  Resistant Phenotype  \n",
       "0  28512093;28776045  573.13031          Susceptible  \n",
       "1  28512093;28776045  573.13032          Susceptible  \n",
       "2  28512093;28776045  573.13033          Susceptible  \n",
       "3  28512093;28776045  573.13034            Resistant  \n",
       "4  28512093;28776045  573.13035          Susceptible  \n",
       "\n",
       "[5 rows x 302 columns]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Shape and first few rows viewing\n",
    "\n",
    "print(\"Shape:\", df.shape)\n",
    "df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "97b9e168-30a9-49c2-85e9-c2cc28bbe74f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['Genome ID',\n",
       " 'AAC(3)-IId',\n",
       " 'AAC(3)-IIe',\n",
       " 'AAC(3)-IV',\n",
       " 'AAC(3)-Ia',\n",
       " \"AAC(6')-29a\",\n",
       " \"AAC(6')-33\",\n",
       " \"AAC(6')-IIa\",\n",
       " \"AAC(6')-IIc\",\n",
       " \"AAC(6')-Ib-cr\",\n",
       " \"AAC(6')-Ib10\",\n",
       " \"AAC(6')-Ib7\",\n",
       " \"AAC(6')-Ib9\",\n",
       " \"AAC(6')-Il\",\n",
       " \"AAC(6')-Im\",\n",
       " \"ANT(2'')-Ia\",\n",
       " \"ANT(3'')-IIa\",\n",
       " \"APH(2'')-IIa\",\n",
       " \"APH(3'')-Ib\",\n",
       " \"APH(3')-Ia\",\n",
       " \"APH(3')-VI\",\n",
       " \"APH(3')-VIa\",\n",
       " 'APH(4)-Ia',\n",
       " 'APH(6)-Id',\n",
       " 'BKC-1',\n",
       " 'CARB-3',\n",
       " 'CMY-59',\n",
       " 'CTX-M-1',\n",
       " 'CTX-M-101',\n",
       " 'CTX-M-14',\n",
       " 'CTX-M-15',\n",
       " 'CTX-M-2',\n",
       " 'CTX-M-27',\n",
       " 'CTX-M-3',\n",
       " 'CTX-M-65',\n",
       " 'CTX-M-9',\n",
       " 'CTX-M-98',\n",
       " 'DHA-1',\n",
       " 'DHA-6',\n",
       " 'EreA2',\n",
       " 'ErmB',\n",
       " 'FOX-5',\n",
       " 'FosA3',\n",
       " 'FosA5',\n",
       " 'FosA6',\n",
       " 'FosA7',\n",
       " 'IMP-14',\n",
       " 'KPC-1',\n",
       " 'KPC-3',\n",
       " 'KPC-8',\n",
       " 'Klebsiella_pneumoniae_KpnE',\n",
       " 'Klebsiella_pneumoniae_KpnF',\n",
       " 'Klebsiella_pneumoniae_KpnG',\n",
       " 'Klebsiella_pneumoniae_OmpK37',\n",
       " 'Klebsiella_pneumoniae_acrA',\n",
       " 'LAP-2',\n",
       " 'LEN-10',\n",
       " 'LEN-16',\n",
       " 'LEN-17',\n",
       " 'LEN-18',\n",
       " 'LEN-2',\n",
       " 'LEN-26',\n",
       " 'LEN-30',\n",
       " 'LEN-55',\n",
       " 'LEN-9',\n",
       " 'MCR-9',\n",
       " 'NDM-1',\n",
       " 'NDM-5',\n",
       " 'OKP-A-3',\n",
       " 'OKP-A-5',\n",
       " 'OKP-A-7',\n",
       " 'OKP-B-19',\n",
       " 'OKP-B-3',\n",
       " 'OKP-B-5',\n",
       " 'OKP-B-6',\n",
       " 'OKP-B-8',\n",
       " 'OXA-1',\n",
       " 'OXA-10',\n",
       " 'OXA-181',\n",
       " 'OXA-2',\n",
       " 'OXA-232',\n",
       " 'OXA-48',\n",
       " 'OXA-9',\n",
       " 'QnrA1',\n",
       " 'QnrA7',\n",
       " 'QnrB17',\n",
       " 'QnrB20',\n",
       " 'QnrB4',\n",
       " 'QnrB5',\n",
       " 'QnrB58',\n",
       " 'QnrS1',\n",
       " 'SAT-1',\n",
       " 'SCO-1',\n",
       " 'SHV-1',\n",
       " 'SHV-100',\n",
       " 'SHV-101',\n",
       " 'SHV-105',\n",
       " 'SHV-106',\n",
       " 'SHV-107',\n",
       " 'SHV-108',\n",
       " 'SHV-11',\n",
       " 'SHV-110',\n",
       " 'SHV-119',\n",
       " 'SHV-12',\n",
       " 'SHV-120',\n",
       " 'SHV-121',\n",
       " 'SHV-134',\n",
       " 'SHV-142',\n",
       " 'SHV-143',\n",
       " 'SHV-155',\n",
       " 'SHV-161',\n",
       " 'SHV-172',\n",
       " 'SHV-178',\n",
       " 'SHV-182',\n",
       " 'SHV-185',\n",
       " 'SHV-187',\n",
       " 'SHV-2',\n",
       " 'SHV-26',\n",
       " 'SHV-27',\n",
       " 'SHV-30',\n",
       " 'SHV-33',\n",
       " 'SHV-36',\n",
       " 'SHV-42',\n",
       " 'SHV-44',\n",
       " 'SHV-5',\n",
       " 'SHV-52',\n",
       " 'SHV-60',\n",
       " 'SHV-65',\n",
       " 'SHV-76',\n",
       " 'SHV-77',\n",
       " 'SHV-86',\n",
       " 'TEM-1',\n",
       " 'TEM-104',\n",
       " 'TEM-12',\n",
       " 'TEM-122',\n",
       " 'TEM-15',\n",
       " 'TEM-150',\n",
       " 'TEM-158',\n",
       " 'TEM-206',\n",
       " 'TEM-79',\n",
       " 'VEB-1',\n",
       " 'VEB-5',\n",
       " 'VIM-19',\n",
       " 'VIM-27',\n",
       " 'aadA13',\n",
       " 'aadA16',\n",
       " 'aadA2',\n",
       " 'aadA5',\n",
       " 'armA',\n",
       " 'arr-2',\n",
       " 'arr-3',\n",
       " 'catB11',\n",
       " 'catB3',\n",
       " 'catI',\n",
       " 'catII_from_Escherichia_coli_K-12',\n",
       " 'cmlA1',\n",
       " 'cmlA4',\n",
       " 'cmlA5',\n",
       " 'cmlA6',\n",
       " 'determinant_of_bleomycin_resistance',\n",
       " 'dfrA1',\n",
       " 'dfrA12',\n",
       " 'dfrA14',\n",
       " 'dfrA15',\n",
       " 'dfrA17',\n",
       " 'dfrA19',\n",
       " 'dfrA22',\n",
       " 'dfrA23',\n",
       " 'dfrA26',\n",
       " 'dfrA27',\n",
       " 'dfrA30',\n",
       " 'dfrA5',\n",
       " 'dfrA7',\n",
       " 'dfrA8',\n",
       " 'dfrB3',\n",
       " 'floR',\n",
       " 'mphA',\n",
       " 'mphE',\n",
       " 'msrE',\n",
       " 'oqxA',\n",
       " 'oqxB',\n",
       " 'qacH',\n",
       " 'rmtB',\n",
       " 'rmtC',\n",
       " 'rmtF',\n",
       " 'sul1',\n",
       " 'sul2',\n",
       " 'sul3',\n",
       " 'tet(A)',\n",
       " 'tet(B)',\n",
       " 'tet(C)',\n",
       " 'tet(D)',\n",
       " 'tet(G)',\n",
       " 'tet(W/N/W)',\n",
       " 'tetM',\n",
       " 'tetW',\n",
       " 'Genome Name_x',\n",
       " 'Other Names',\n",
       " 'NCBI Taxon ID',\n",
       " 'Taxon Lineage IDs',\n",
       " 'Taxon Lineage Names',\n",
       " 'Superkingdom',\n",
       " 'Kingdom',\n",
       " 'Phylum',\n",
       " 'Class',\n",
       " 'Order',\n",
       " 'Family',\n",
       " 'Genus',\n",
       " 'Species',\n",
       " 'Genome Status',\n",
       " 'Strain',\n",
       " 'Serovar',\n",
       " 'Biovar',\n",
       " 'Pathovar',\n",
       " 'MLST',\n",
       " 'Segment',\n",
       " 'Subtype',\n",
       " 'H_type',\n",
       " 'N_type',\n",
       " 'H1 Clade Global',\n",
       " 'H1 Clade US',\n",
       " 'H5 Clade',\n",
       " 'pH1N1-like',\n",
       " 'Lineage',\n",
       " 'Clade',\n",
       " 'Subclade',\n",
       " 'Other Typing',\n",
       " 'Culture Collection',\n",
       " 'Type Strain',\n",
       " 'Reference',\n",
       " 'Genome Quality',\n",
       " 'Completion Date',\n",
       " 'Publication',\n",
       " 'Authors',\n",
       " 'BioProject Accession',\n",
       " 'BioSample Accession',\n",
       " 'SRA Accession',\n",
       " 'GenBank Accessions',\n",
       " 'Sequencing Center',\n",
       " 'Sequencing Status',\n",
       " 'Sequencing Platform',\n",
       " 'Sequencing Depth',\n",
       " 'Assembly Method',\n",
       " 'Chromosome',\n",
       " 'Plasmids',\n",
       " 'Contigs',\n",
       " 'Size',\n",
       " 'GC Content',\n",
       " 'Contig L50',\n",
       " 'Contig N50',\n",
       " 'TRNA',\n",
       " 'RRNA',\n",
       " 'Mat Peptide',\n",
       " 'CDS',\n",
       " 'Coarse Consistency',\n",
       " 'Fine Consistency',\n",
       " 'CheckM Contamination',\n",
       " 'CheckM Completeness',\n",
       " 'Genome Quality Flags',\n",
       " 'Isolation Source',\n",
       " 'Isolation Comments',\n",
       " 'Collection Date',\n",
       " 'Collection Year',\n",
       " 'Season',\n",
       " 'Isolation Country',\n",
       " 'Geographic Group',\n",
       " 'Geographic Location',\n",
       " 'Other Environmental',\n",
       " 'Host Name',\n",
       " 'Host Common Name',\n",
       " 'Host Gender',\n",
       " 'Host Age',\n",
       " 'Host Health',\n",
       " 'Host Group',\n",
       " 'Lab Host',\n",
       " 'Passage',\n",
       " 'Other Clinical',\n",
       " 'Additional Metadata',\n",
       " 'Comments',\n",
       " 'Date Inserted',\n",
       " 'Date Modified',\n",
       " 'Taxon ID',\n",
       " 'Genome Name_y',\n",
       " 'Antibiotic',\n",
       " 'Measurement',\n",
       " 'Measurement Sign',\n",
       " 'Measurement Value',\n",
       " 'Measurement Unit',\n",
       " 'Laboratory Typing Method',\n",
       " 'Laboratory Typing Method Version',\n",
       " 'Laboratory Typing Platform',\n",
       " 'Vendor',\n",
       " 'Testing Standard',\n",
       " 'Testing Standard Year',\n",
       " 'Computational Method',\n",
       " 'Computational Method Version',\n",
       " 'Computational Method Performance',\n",
       " 'Evidence',\n",
       " 'Source',\n",
       " 'PubMed',\n",
       " 'SEQUENCE',\n",
       " 'Resistant Phenotype']"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#List of all columns\n",
    "\n",
    "df.columns.tolist()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e5cacb47-fe94-4cb5-aecf-162befd92f4e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Resistant Phenotype\n",
       "Susceptible    1054\n",
       "Resistant       526\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Check Resistant Values\n",
    "df[\"Resistant Phenotype\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "cf7f3700-2a39-430e-b0eb-e959e5df5e4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Convert Label Names for y value\n",
    "\n",
    "y = df[\"Resistant Phenotype\"].map({\"Susceptible\": 0, \"Resistant\": 1})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "496012de-84e6-44ce-991e-b307efb8d516",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Eliminate Non-genetic metadata \n",
    "\n",
    "drop_cols = ['Resistant Phenotype', 'Genome ID', 'Genome Name_x', 'Genome Name_y',\n",
    "    'Antibiotic', 'Measurement', 'Measurement Sign', 'Measurement Value', 'Measurement Unit',\n",
    "    'Testing Standard', 'Testing Standard Year', 'Computational Method', 'Computational Method Version',\n",
    "    'Vendor', 'Evidence', 'Source', 'PubMed', 'SEQUENCE', 'Date Inserted', 'Date Modified',\n",
    "    'BioProject Accession', 'BioSample Accession', 'SRA Accession', 'Host Name', 'Collection Date',\n",
    "    'Size', 'Contigs', 'Plasmids', 'Genome Quality', 'Taxon ID', 'Species', 'Strain']\n",
    "\n",
    "#x value\n",
    "\n",
    "x = df.drop(columns= drop_cols, errors= \"ignore\")\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2b263bc7-ee66-4461-99d4-9cc13d6e1998",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'C:\\\\Users\\\\eisan\\\\PycharmProjects\\\\AMR first\\\\notebooks'"
      ]
     },
     "execution_count": 1,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "783f06d9-ac4e-4d2d-9c5b-6f4e23b21cf0",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
